tutorial/tests/test_object_oriented_programming.py

```diff
@@ -2,6 +2,202 @@
 
 import pytest
 
+#
+# Example 1: Person
+#
+
+
+def reference_oop_person(first_name: str, last_name: str):
+    class Person:
+        """A class representing a person with first name and last name"""
+
+        def __init__(self, first_name: str, last_name: str):
+            self.first_name = first_name
+            self.last_name = last_name
+
+    return Person(first_name, last_name)
+
+
+def validate_oop_person(solution_result):
+    params = list(vars(solution_result))
+    assert (
+        type(solution_result).__name__ == "Person"
+    ), "The class should be named 'Person'."
+    assert len(params) == 2, "The class should have 2 attributes."
+    assert (
+        "first_name" in params and "last_name" in params
+    ), "The class attributes should be 'first_name' and 'last_name'."
+
+
+@pytest.mark.parametrize(
+    "first_name, last_name",
+    [
+        ("Jane", "Doe"),
+        ("John", "Doe"),
+    ],
+)
+def test_oop_person(first_name, last_name, function_to_test):
+    solution_result = function_to_test(first_name, last_name)
+    reference_result = reference_oop_person(first_name, last_name)
+
+    validate_oop_person(solution_result)
+    assert (
+        solution_result.first_name == reference_result.first_name
+        and solution_result.last_name == reference_result.last_name
+    )
+
+
+#
+# Example 2: Person's full name
+#
+
+
+def reference_oop_fullname(first_name: str, last_name: str):
+    class Person:
+        """A class representing a person with first name and last name"""
+
+        def __init__(self, first_name: str, last_name: str):
+            self.first_name = first_name
+            self.last_name = last_name
+
+        def full_name(self) -> str:
+            return f"My name is {self.first_name} {self.last_name}"
+
+    return Person(first_name, last_name)
+
+
+def validate_oop_fullname(solution_result):
+    methods = [
+        attr
+        for attr in dir(solution_result)
+        if callable(getattr(solution_result, attr)) and not attr.startswith("__")
+    ]
+    assert len(methods) == 1, "The class should have 1 method."
+    assert "full_name" in methods, "The class method should be called 'full_name'."
+
+
+@pytest.mark.parametrize(
+    "first_name, last_name",
+    [
+        ("Jane", "Doe"),
+        ("John", "Doe"),
+    ],
+)
+def test_oop_fullname(first_name, last_name, function_to_test):

```

Comment:
Since the exercise says fullname should return My name is {first_name} {last_name}, I can imagine some people will do this which also passes the test but is a difficult mistake to spot in general so I would test that here, below i made a suggestion.

Reply:
Ahh, or is it that they will use `first_name` instead of `self.first_name`?